dbt/include/athena/macros/materializations/models/incremental/merge.sql

```diff
@@ -96,15 +96,17 @@
           when matched and ({{ delete_condition }})
           then delete
       {%- endif %}
-      when matched
-        then update set
-          {%- for col in update_columns %}
-            {%- if merge_update_columns_rules and col.name in merge_update_columns_rules %}
-              {{ get_update_statement(col, merge_update_columns_rules[col.name], loop.last) }}
-            {%- else -%}
-              {{ get_update_statement(col, merge_update_columns_default_rule, loop.last) }}
-            {%- endif -%}
-          {%- endfor %}
+      {% if update_columns -%}
```

Comment:
does it mean that we have to put all `update_columns` in `merge_update_columns ` if so this is not really convienent

Reply:
Sorry, could you elaborate? The fix doesn't change handling of `update_columns`, the way is constructed. Only if it's completely empty, the update part is skipped to allow insert only merges and avoid generating invalid merge statement.

Reply:
Ok the key is in https://github.com/dbt-athena/dbt-athena/blob/main/dbt/include/athena/macros/materializations/models/incremental/merge.sql#L1-L27 - if merge_update_columns is not set and merge_exclude_columns is not set, the macro fallsback to all destination columns making a full update on all the columns. 

I believe that might be worth to document such new behvior and how can people use it.
e.g. if I don't have to have updates the user must set `merge_update_columns` as empty array right? that should result in update_columns being null and nothing to update.

Reply:
> Ok the key is in https://github.com/dbt-athena/dbt-athena/blob/main/dbt/include/athena/macros/materializations/models/incremental/merge.sql#L1-L27 - if merge_update_columns is not set and merge_exclude_columns is not set, the macro fallsback to all destination columns making a full update on all the columns.

Ok, and in case, `update_columns` has all columns, it's still should work as before, right? Unless I completely miss something.

>  if I don't have to have updates the user must set merge_update_columns as empty array right?
No, the update is skipped only if the final list of update columns is empty. At least, for the original case it can be triggered by limiting `merge_update_columns` to `unique_key`. Maybe there are other cases that can result in an empty update_columns. But in any case, it's better to handle it than generate invalid sql, imho.

Reply:
> No, the update is skipped only if the final list of update columns is empty. At least, for the original case it can be triggered by limiting merge_update_columns to unique_key. Maybe there are other cases that can result in an empty update_columns. But in any case, it's better to handle it than generate invalid sql, IMHO.

ah got it, It's a sort of a trick: if `merge_update_columns=['id']` then the update will be still executed because merge_update_columns is not null.

I still don't get how get_merge_update_columns can return no columns. Pretty much seems that if get_merge_update_columns receive an empty merge_update_columns and merge_exclude_columns will fallback to all columns.
But if we pass merge_update_columns=['id'] I will return ['id'], leading to have your new condition evaluate to True.

I still believe that what you propose is not enough ;)

For example you can tweak get_merge_update_columns in a way that if unique_key=merge_update_columns, the it returns no columns, only then your condition will be evaluated to False, an finally the update skipped.

Still to avoid unexpected behaviors and simplify reviewers' lives:
* Provide models used to test - even if they come from elementary, extract them and put them on top of the PR
* add functional tests - MERGE is such a crucial feature with iceberg tables that we cannot risk breaking it ;)

Reply:
@artem-garmash did you give another thought at my message?

I believe that we could simply add something like this somewhere
```
{% if unique_key==merge_update_columns -%} 
    {%- set update_columns = [] -%} 
{% endif -%} 
```
that should cover exactly what you want to achieve, therefore `{% if update_columns -%}` will be evaluated properly.

Reply:
@nicor88 , what is exactly wrong with using `{% if update_columns -%}`? To me it looks a more general fix as it prevents generating invalid sql regardless of how update_columns can be empty. I updated the description and tests, to make it more clear.

Reply:
@nicor88 actually I don't see a problem with @artem-garmash solution. Do I miss any corner cases ?

Reply:
@artem-garmash there is no problem with adding `{% if update_columns -%}` totally fine. 
Not sure if I explained my point properly, in case ```unique_key==merge_update_columns``` update_columns must be empty and it's not the case for the current implementation, therefore we might want to added that corner case somewhere, otherwise update_columns will be never empty and we always have it in the merge statement.
Not sure if it's more clear now ;)

Reply:
as I see we get update_columns from columns list without unique_key columns. That's why we can get empty list in case when unique_key == merge_update_columns. That's why I see the condition offered by @artem-garmash looks elegant to me. 
```
{%- set update_columns = get_merge_update_columns(merge_update_columns, merge_exclude_columns, dest_columns_wo_keys) -%}
```

>Not sure if I explained my point properly, in case unique_key==merge_update_columns update_columns should be empty (but it won't be)

why won't it be empty ?

Reply:
Maybe I'm missing something but get_merge_update_columns will never return empty columns.

can you give me a practical example when get_merge_update_columns will return an emtpy columns, providing an explicative model configuration?

Specifically here some thoghts:
* merge_update_columns is not empty -> get_merge_update_columns will return merge_update_columns
* merge_exclude_columns is not empty -> get_merge_update_columns dest_columns minus merge_exclude_columns
* merge_update_columns and merge_exclude_columns are empty -> get_merge_update_columns will returns dest_columns

This means that if merge_update_columns is empty we always fallback to the all destination columns.

Reply:
> Maybe I'm missing something but `get_merge_update_columns()` will never return empty columns.

It does return empty list, the 2 cases in the PR fail without the check for empty update_columns (see models in the description):

1. For the `merge_update_columns` == `unique_id` case, it will return empty list cause `dest_columns` will not have `unique_id` (as `unique_id` is removed in passed `dest_columns_wo_keys` in the call). The function goes to the `{%- elif merge_update_columns -%}` branch and the if in the loop will be skipped as `merge_exclude_columns` contains only removed `unique_id`, it basically intersection between `dest_columns` and `merge_exclude_columns`. In the first test case it would be `dest_columns = ["msg", "color"], merge_exclude_columns=["id"]` - no common columns.
2. For the `merge_exclude_columns` case, it goes to the `{%- elif merge_exclude_columns -%}` branch and in a similar way `merge_exclude_columns` may have all destination columns and the result is empty. So for the test case, it would be `dest_columns = ["msg", "color"], merge_exclude_columns=["msg", "color"]` and the if is negative, so it's difference and it's empty again.

The `{%- set update_columns = dest_columns -%}` branch you are referring to is only invoked if both `merge_update_columns` and `merge_exclude_columns` are empty, right?

Of course, another way to fix this is to remove removing of uniqe_id, to guarantee update_columns is never empty.

I'm looking at it just at sql construction level, the update part in the merge query should never be added is the update list is empty, otherwise is invalid sql. So it works regardless of changes in `get_merge_update_columns()`.